File: go-cgo-impl/registry.go
package main

// OrderRegistry holds flat per-order metadata indexed directly by OrderID.
// Access is O(1) by array index, avoiding map overhead on the hot read path.
type OrderRegistry struct {
	Meta [NumOrders]OrderMeta // indexed by order_id; order_id must be < NumOrders
}

// NewOrderRegistry allocates and returns an empty OrderRegistry.
// All slots are zero-valued (Active=false, PBTDeadline=0).
func NewOrderRegistry() *OrderRegistry {
	return &OrderRegistry{}
}

// Upsert inserts or updates the metadata for an order.
// The order's ID must be < NumOrders; orders outside this range are silently dropped.
func (r *OrderRegistry) Upsert(m OrderMeta) {
	if m.OrderID >= NumOrders {
		return
	}
	r.Meta[m.OrderID] = m
}

// Deactivate marks the order as inactive so it is excluded from future ranking cycles.
func (r *OrderRegistry) Deactivate(orderID uint32) {
	if orderID >= NumOrders {
		return
	}
	r.Meta[orderID].Active = false
}

// EvictOrder deactivates an order and cleans up its index entries and PPSMatrix slots.
// Use this for Kafka order-removed events.
func (r *OrderRegistry) EvictOrder(orderID uint32, indexes *InvertedIndexes, matrix *PPSMatrix) {
	if orderID >= NumOrders || !r.Meta[orderID].Active {
		return
	}
	r.Meta[orderID].Active = false
	r.Meta[orderID].InsertedAtSecs = 0
	indexes.RemoveOrder(orderID)
	matrix.EvictOrder(orderID)
}

// PurgeExpired scans all orders and evicts those whose TTL has expired.
func (r *OrderRegistry) PurgeExpired(nowSecs int64, indexes *InvertedIndexes, matrix *PPSMatrix) {
	for i := 0; i < NumOrders; i++ {
		m := &r.Meta[i]
		if m.Active && m.InsertedAtSecs > 0 && (nowSecs-m.InsertedAtSecs) > OrderTTLSecs {
			orderID := m.OrderID
			m.Active = false
			m.InsertedAtSecs = 0
			indexes.RemoveOrder(orderID)
			matrix.EvictOrder(orderID)
		}
	}
}

// InvertedIndexes maps TPID and BinTag IDs to sets of order IDs.
// Includes reverse maps for O(1) per-order cleanup during eviction.
type InvertedIndexes struct {
	TPIDToOrders   map[uint64][]uint32 // TPID → order IDs
	BinTagToOrders map[uint32][]uint32 // BinTag → eligible order IDs
	reverseTPID    map[uint32][]uint64 // orderID → TPIDs it's registered under
	reverseBinTag  map[uint32][]uint32 // orderID → BinTag IDs it's registered under
}

// NewInvertedIndexes allocates and returns empty inverted indexes.
func NewInvertedIndexes() *InvertedIndexes {
	return &InvertedIndexes{
		TPIDToOrders:   make(map[uint64][]uint32),
		BinTagToOrders: make(map[uint32][]uint32),
		reverseTPID:    make(map[uint32][]uint64),
		reverseBinTag:  make(map[uint32][]uint32),
	}
}

// AddOrder registers orderID under all given TPIDs and BinTag IDs (idempotent upsert).
// Callers should call RemoveOrder before re-adding if an order is being updated.
func (ix *InvertedIndexes) AddOrder(orderID uint32, tpids []uint64, bintagIDs []uint32) {
	ix.RemoveOrder(orderID) // idempotent: remove stale entries first

	for _, t := range tpids {
		ix.TPIDToOrders[t] = append(ix.TPIDToOrders[t], orderID)
	}
	for _, b := range bintagIDs {
		ix.BinTagToOrders[b] = append(ix.BinTagToOrders[b], orderID)
	}

	if len(tpids) > 0 {
		cp := make([]uint64, len(tpids))
		copy(cp, tpids)
		ix.reverseTPID[orderID] = cp
	}
	if len(bintagIDs) > 0 {
		cp := make([]uint32, len(bintagIDs))
		copy(cp, bintagIDs)
		ix.reverseBinTag[orderID] = cp
	}
}

// RemoveOrder removes all index entries for orderID using the reverse maps.
func (ix *InvertedIndexes) RemoveOrder(orderID uint32) {
	for _, tpid := range ix.reverseTPID[orderID] {
		list := ix.TPIDToOrders[tpid]
		list = removeUint32(list, orderID)
		if len(list) == 0 {
			delete(ix.TPIDToOrders, tpid)
		} else {
			ix.TPIDToOrders[tpid] = list
		}
	}
	delete(ix.reverseTPID, orderID)

	for _, btID := range ix.reverseBinTag[orderID] {
		list := ix.BinTagToOrders[btID]
		list = removeUint32(list, orderID)
		if len(list) == 0 {
			delete(ix.BinTagToOrders, btID)
		} else {
			ix.BinTagToOrders[btID] = list
		}
	}
	delete(ix.reverseBinTag, orderID)
}

// removeUint32 removes the first occurrence of v from s (swap-and-shrink).
func removeUint32(s []uint32, v uint32) []uint32 {
	for i, x := range s {
		if x == v {
			return append(s[:i], s[i+1:]...)
		}
	}
	return s
}

File: java_kotlin/src/main/kotlin/com/greyorange/ps/Registry.kt
package com.greyorange.ps

// ── OrderRegistry ────────────────────────────────────────────────────────────
// Struct-of-Arrays layout: parallel primitive arrays indexed by orderId.
//
// WHY SoA and not Array<OrderMeta>:
//   JVM object headers are 16 bytes each. Array<OrderMeta> with 1M entries =
//   1M × (16B header + 4+4+8+1+8 = 25B fields + 7B padding) ≈ 52 MB + pointer
//   array overhead. SoA gives 4+4+8+1+8 = 25 MB flat — matching Go's layout.
//   Phase 2 access pattern (active check + qty + pbtDeadline) hits 3 separate
//   arrays; Go hits 3 fields in one struct. Cache behaviour is approximately
//   equal at 1M scale — the benchmark will show the real difference.

class OrderRegistry {
    // All arrays are indexed by orderId (0-based, orderId < NUM_ORDERS).
    val active         = BooleanArray(NUM_ORDERS)
    val requiredQty    = FloatArray(NUM_ORDERS)
    val pbtDeadline    = LongArray(NUM_ORDERS)      // Unix seconds; -1 = no deadline
    val insertedAtSecs = LongArray(NUM_ORDERS)       // for TTL eviction

    fun upsert(
        orderId: Int,
        qty: Float,
        pbtDl: Long,
        insertedAt: Long,
    ) {
        if (orderId >= NUM_ORDERS) return
        active[orderId]         = true
        requiredQty[orderId]    = qty
        pbtDeadline[orderId]    = pbtDl
        insertedAtSecs[orderId] = insertedAt
    }

    fun deactivate(orderId: Int) {
        if (orderId >= NUM_ORDERS) return
        active[orderId]         = false
        insertedAtSecs[orderId] = 0
    }

    fun evictOrder(orderId: Int, indexes: InvertedIndexes, matrix: PPSMatrix) {
        if (orderId >= NUM_ORDERS || !active[orderId]) return
        active[orderId]         = false
        insertedAtSecs[orderId] = 0
        indexes.removeOrder(orderId)
        matrix.evictOrder(orderId)
    }

    fun purgeExpired(nowSecs: Long, indexes: InvertedIndexes, matrix: PPSMatrix) {
        for (i in 0 until NUM_ORDERS) {
            val ins = insertedAtSecs[i]
            if (active[i] && ins > 0 && (nowSecs - ins) > ORDER_TTL_SECS) {
                active[i]         = false
                insertedAtSecs[i] = 0
                indexes.removeOrder(i)
                matrix.evictOrder(i)
            }
        }
    }

    fun activeCount(): Int {
        var n = 0
        for (i in 0 until NUM_ORDERS) if (active[i]) n++
        return n
    }
}

// ── InvertedIndexes ──────────────────────────────────────────────────────────
// Same dual-index structure as Go:
//   tpidToOrders    : Map<tpid: Long, List<orderId: Int>>
//   binTagToOrders  : Map<binTagId: Int, List<orderId: Int>>
// Reverse maps for O(1) cleanup on eviction.

class InvertedIndexes {
    val tpidToOrders   = HashMap<Long, MutableList<Int>>()    // tpid → [orderId...]
    val binTagToOrders = HashMap<Int,  MutableList<Int>>()    // binTagId → [orderId...]

    // reverse maps for O(1) per-order cleanup
    private val revTpid   = HashMap<Int, MutableList<Long>>() // orderId → [tpid...]
    private val revBinTag = HashMap<Int, MutableList<Int>>()  // orderId → [binTagId...]

    fun addOrder(orderId: Int, tpids: LongArray, binTagIds: IntArray) {
        removeOrder(orderId)  // idempotent: clear stale entries first

        for (t in tpids) tpidToOrders.getOrPut(t) { mutableListOf() }.add(orderId)
        for (b in binTagIds) binTagToOrders.getOrPut(b) { mutableListOf() }.add(orderId)

        if (tpids.isNotEmpty())     revTpid[orderId]   = tpids.toMutableList()
        if (binTagIds.isNotEmpty()) revBinTag[orderId]  = binTagIds.toMutableList()
    }

    // ── Materialized (frozen) snapshots for hot-path loops ───────────────────
    // MutableList<Int> stores boxed java.lang.Integer. Every forEach on these lists
    // in Phase 1 / Phase 2 causes N unbox operations per cycle.
    // Call materialize() + materializeTpid() once after bulk ingestion; the frozen
    // IntArrays iterate with zero boxing — identical to Go's []uint32 / Rust's Vec<u32>.

    var binTagFrozen: HashMap<Int,  IntArray> = HashMap(0)
        private set

    var tpidFrozen:   HashMap<Long, IntArray> = HashMap(0)
        private set

    fun materialize() {
        binTagFrozen = freezeIntLists(binTagToOrders)
    }

    fun materializeTpid() {
        tpidFrozen = freezeLongLists(tpidToOrders)
    }

    private fun freezeIntLists(src: HashMap<Int, MutableList<Int>>): HashMap<Int, IntArray> {
        val snap = HashMap<Int, IntArray>(src.size * 2)
        for ((k, list) in src) {
            val arr = IntArray(list.size); var i = 0
            for (v in list) arr[i++] = v
            snap[k] = arr
        }
        return snap
    }

    private fun freezeLongLists(src: HashMap<Long, MutableList<Int>>): HashMap<Long, IntArray> {
        val snap = HashMap<Long, IntArray>(src.size * 2)
        for ((k, list) in src) {
            val arr = IntArray(list.size); var i = 0
            for (v in list) arr[i++] = v
            snap[k] = arr
        }
        return snap
    }

    fun removeOrder(orderId: Int) {
        revTpid.remove(orderId)?.forEach { t ->
            val list = tpidToOrders[t] ?: return@forEach
            list.remove(orderId)
            if (list.isEmpty()) tpidToOrders.remove(t)
        }
        revBinTag.remove(orderId)?.forEach { b ->
            val list = binTagToOrders[b] ?: return@forEach
            list.remove(orderId)
            if (list.isEmpty()) binTagToOrders.remove(b)
        }
    }
}

// Extension to convert LongArray to MutableList<Long> cleanly.
private fun LongArray.toMutableList(): MutableList<Long> =
    ArrayList<Long>(this.size).also { for (v in this) it.add(v) }
